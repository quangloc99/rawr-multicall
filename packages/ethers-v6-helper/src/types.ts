export type MethodNames<C> = C extends {
    interface: {
        getFunction(name: infer Name extends string, ...params: unknown[]): unknown;
    };
}
    ? Name
    : string;

// TODO strip the overrides
export type MethodParameters<C, Name extends MethodNames<C>> = C extends {
    [key in Name]: {
        staticCall(...params: infer NamedParams): unknown;
    };
}
    ? NamedParams
    : unknown[];

export type MethodReturnType<C, Name extends MethodNames<C>> = C extends {
    [key in Name]: {
        staticCall(...params: unknown[]): infer ReturnType;
    };
}
    ? Awaited<ReturnType>
    : unknown;
