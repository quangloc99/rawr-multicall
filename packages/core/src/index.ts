export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './Bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './errors';
export * from './registerPredeployContract';
export * from './Result';
