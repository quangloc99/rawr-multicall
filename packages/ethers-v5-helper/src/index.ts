export * from './types';
export * from './createEthersV5Call';
export * from './error';
