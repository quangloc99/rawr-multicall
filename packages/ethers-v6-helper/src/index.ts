export * from './types';
export * from './error';
export * from './createEthersV6Call';
