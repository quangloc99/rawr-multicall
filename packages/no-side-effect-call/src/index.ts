import { registerPredeployCallThenRevert } from './registerPredeployCallThenRevert';
registerPredeployCallThenRevert();

export * from './CallThenRevertContractArtifact';
export * from './registerPredeployCallThenRevert';
export * from './wrapNoSideEffectCall';
