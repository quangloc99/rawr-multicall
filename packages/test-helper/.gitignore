artifacts
ethers-v6-contracts/typechain-types