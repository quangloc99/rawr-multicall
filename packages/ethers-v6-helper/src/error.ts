import { Bytes, RawMulticallError, bytesToHexWith0x } from '@raw-multicall/core';
import { ErrorFragment } from 'ethers';

export class NoFragmentFoundError extends RawMulticallError {
    constructor(
        readonly data: Bytes,
        message: string = `No error fragment found for reverted data "${bytesToHexWith0x(data)}"`,
        options?: ErrorOptions
    ) {
        super(message, options);
    }
}

export class EthersV6ContractError extends RawMulticallError {
    constructor(
        readonly fragment: ErrorFragment,
        readonly decodedParams: unknown[],
        readonly data: Bytes,
        options?: ErrorOptions
    ) {
        super(`${fragment.name}(${EthersV6ContractError.joinParams(decodedParams)})`, options);
    }

    static joinParams(params: unknown[]) {
        return params.map((value) => (typeof value == 'bigint' ? value : JSON.stringify(value))).join(', ');
    }
}
