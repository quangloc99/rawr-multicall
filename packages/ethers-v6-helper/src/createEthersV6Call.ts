import { AddressOrRawAddress, Call, strip0x, add0x, castToAddress } from '@raw-multicall/core';
import { BaseContract } from 'ethers';
import { MethodNames, MethodParameters, MethodReturnType } from './types';
import { NoFragmentFoundError } from './error';

export type CreateEthersV6CallParams = {
    withAddress?: AddressOrRawAddress;
};

export async function createEthersV6Call<
    C extends BaseContract,
    Method extends MethodNames<C>,
    MethodParams = MethodParameters<C, Method>,
    ReturnType = MethodReturnType<C, Method>,
>(
    contract: C,
    methodName: Method,
    methodParams: MethodParams,
    params: CreateEthersV6CallParams = {}
): Promise<Call<ReturnType, unknown>> {
    const address = castToAddress(params.withAddress != undefined ? params.withAddress : await contract.getAddress());
    // eslint-disable-next-line
    const data = (await (contract as any)[methodName].populateTransaction(...(methodParams as any))).data as string;
    return {
        getContractAddress: () => address,
        getData: () => data,
        decodeResult(data): ReturnType {
            const res = contract.interface.decodeFunctionResult(methodName, data);
            if (res.length == 1) return res[0] as ReturnType;
            return res as ReturnType;
        },
        decodeError(data): unknown {
            try {
                const fragment = contract.interface.getError(add0x(strip0x(data).slice(0, 8)));
                if (!fragment) return new NoFragmentFoundError(data);
                return contract.interface.decodeErrorResult(fragment, add0x(data));
            } catch (e) {
                return e;
            }
        },
    };
}
