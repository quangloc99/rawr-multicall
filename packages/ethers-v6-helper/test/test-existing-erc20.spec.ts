import { createEthersV6Call } from '../src';
import { buildRawMulticallContract, decodeResult, Bytes } from '@raw-multicall/core';
import { CHAIN_ID_MAPPING, describeForChain } from '@raw-multicall/test-helper';
import { BaseContract, JsonRpcProvider } from 'ethers';
import { ERC20, ERC20__factory } from './typechain-types';

const TEST_DATA: Record<
    number,
    {
        blockNumber: number;
        tokenAddresses: string[];
        holders: string[];
    }
> = {
    [CHAIN_ID_MAPPING.ETHEREUM]: {
        blockNumber: 17670778,
        tokenAddresses: [
            '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', // usdc
            '0x808507121b80c02388fad14726482e061b8da827', // pendle
        ],
        holders: ['0xa3a7b6f88361f48403514059f1f16c8e78d60eec', '0x28c6c06298d514db089934071355e5743bf21d60'],
    },
    [CHAIN_ID_MAPPING.ARBITRUM]: {
        blockNumber: 110135034,
        tokenAddresses: [
            '0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9', // usdt,
            '0x0c880f6761F1af8d9Aa9C466984b80DAb9a8c9e8', // pendle
        ],
        holders: ['0xb38e8c17e38363af6ebdcb3dae12e0243582891d', '0xf89d7b9c864f589bbf53a82105107622b35eaa40'],
    },
};
describeForChain(
    (chain: number) => `buildRawMulticallContract for chain ${chain}`,
    (rpcUrl: string, chain: number) => {
        const CUR_TEST_DATA = TEST_DATA[chain];

        let provider: JsonRpcProvider;
        const blockTag = CUR_TEST_DATA.blockNumber;

        const doSend = (data: Bytes) => provider.call({ data, blockTag });
        const allowPUSH0 = chain != CHAIN_ID_MAPPING.ARBITRUM;

        beforeAll(() => {
            provider = new JsonRpcProvider(rpcUrl);
        });

        describe('erc20', () => {
            it('simple', async () => {
                const contract = new BaseContract(CUR_TEST_DATA.tokenAddresses[0], ERC20__factory.abi) as ERC20;
                const calls = [await createEthersV6Call(contract, 'name', [])];
                console.log(calls[0].getContractAddress());
                console.log(calls[0].getData());
                const sendData = buildRawMulticallContract(calls, { allowPUSH0 });
                expect(sendData).toMatchSnapshot();
                const res = await doSend(sendData.byteCode);
                expect(res).toMatchSnapshot();
                const output = decodeResult(calls, res);
                expect(output).toMatchSnapshot();
            });

            it('complex', async () => {
                const calls = await Promise.all(
                    CUR_TEST_DATA.tokenAddresses.map((tokenAddress) => {
                        const contract = new BaseContract(tokenAddress, ERC20__factory.abi) as ERC20;

                        return Promise.all([
                            createEthersV6Call(contract, 'name', []),
                            createEthersV6Call(contract, 'symbol', []),
                            createEthersV6Call(contract, 'decimals', []),
                            createEthersV6Call(contract, 'totalSupply', []),
                            ...CUR_TEST_DATA.holders.map((holder) =>
                                createEthersV6Call(contract, 'balanceOf', [holder])
                            ),
                        ] as const);
                    })
                ).then((res) => res.flat());

                const sendData = buildRawMulticallContract(calls, { allowPUSH0 });
                expect(sendData).toMatchSnapshot();
                const res = await doSend(sendData.byteCode);
                expect(res).toMatchSnapshot();
                const output = decodeResult(calls, res);
                expect(output).toMatchSnapshot();
            });
        });
    }
);
