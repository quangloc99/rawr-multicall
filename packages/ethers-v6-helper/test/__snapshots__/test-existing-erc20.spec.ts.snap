// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildRawMulticallContract for chain 1 erc20 complex 1`] = `
{
  "byteCode": "0x60586200024060803960ff60e0600460d85f5f600460805f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460845f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460885f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f6004608c5f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f602460905f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f602460b45f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460805f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f600460845f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f600460885f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f6004608c5f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f602460905f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f602460b45f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d0160d8900360d8f306fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec70a0823100000000000000000000000028c6c06298d514db089934071355e5743bf21d60",
  "splittedByteCodes": [
    "6058",
    "62000240",
    "6080",
    "39",
    "60ff",
    "60e0",
    "6004",
    "60d8",
    "5f",
    "5f",
    "6004",
    "6080",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "6084",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "6088",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "608c",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6024",
    "6090",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6024",
    "60b4",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "6080",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "6084",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "6088",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6004",
    "608c",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6024",
    "6090",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6024",
    "60b4",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "60d8",
    "90",
    "03",
    "60d8",
    "f3",
    "",
    "",
    "06fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec70a0823100000000000000000000000028c6c06298d514db089934071355e5743bf21d60",
  ],
}
`;

exports[`buildRawMulticallContract for chain 1 erc20 complex 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000855534420436f696e0000000000000000000000000000000000000000000000008000006000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000004555344430000000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000000680000020000000000000000000000000000000000000000000000000005a9afec6a2aa60800000200000000000000000000000000000000000000000000000000000000000000000800000200000000000000000000000000000000000000000000000000000946e5d46833c800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c45000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000001280000020000000000000000000000000000000000000000000bfadbf953a8e088d5e9ce68000002000000000000000000000000000000000000000000011009602009976b7403ba5800000200000000000000000000000000000000000000000000d331ff1ff3cdae19b2127"`;

exports[`buildRawMulticallContract for chain 1 erc20 complex 3`] = `
[
  {
    "result": "USD Coin",
    "success": true,
  },
  {
    "result": "USDC",
    "success": true,
  },
  {
    "result": 6n,
    "success": true,
  },
  {
    "result": 25503166948878944n,
    "success": true,
  },
  {
    "result": 0n,
    "success": true,
  },
  {
    "result": 163201732215612n,
    "success": true,
  },
  {
    "result": "Pendle",
    "success": true,
  },
  {
    "result": "PENDLE",
    "success": true,
  },
  {
    "result": 18n,
    "success": true,
  },
  {
    "result": 231725335029111927198358758n,
    "success": true,
  },
  {
    "result": 20554506089343675638234021n,
    "success": true,
  },
  {
    "result": 15957465632406009160605991n,
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 1`] = `
{
  "byteCode": "0x60046200004660803960ff60e0600460845f5f600460805f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d01608490036084f306fdde03",
  "splittedByteCodes": [
    "6004",
    "62000046",
    "6080",
    "39",
    "60ff",
    "60e0",
    "6004",
    "6084",
    "5f",
    "5f",
    "6004",
    "6080",
    "5f",
    "73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "6084",
    "90",
    "03",
    "6084",
    "f3",
    "",
    "",
    "06fdde03",
  ],
}
`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000855534420436f696e000000000000000000000000000000000000000000000000"`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 3`] = `
[
  {
    "result": "USD Coin",
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 1`] = `
{
  "byteCode": "0x60586200027060803960ff60e0600460d86000600060046080600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060046084600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060046088600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d01600060006004608c600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060246090600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d0160006000602460b4600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d0160006000600460806000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000600460846000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000600460886000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d01600060006004608c6000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000602460906000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000602460b46000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160d8900360d8f306fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d70a08231000000000000000000000000f89d7b9c864f589bbf53a82105107622b35eaa40",
  "splittedByteCodes": [
    "6058",
    "62000270",
    "6080",
    "39",
    "60ff",
    "60e0",
    "6004",
    "60d8",
    "6000",
    "6000",
    "6004",
    "6080",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "6084",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "6088",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "608c",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6024",
    "6090",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6024",
    "60b4",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "6080",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "6084",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "6088",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6004",
    "608c",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6024",
    "6090",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6024",
    "60b4",
    "6000",
    "730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "60d8",
    "90",
    "03",
    "60d8",
    "f3",
    "",
    "",
    "06fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d70a08231000000000000000000000000f89d7b9c864f589bbf53a82105107622b35eaa40",
  ],
}
`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000a546574686572205553440000000000000000000000000000000000000000000080000060000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000045553445400000000000000000000000000000000000000000000000000000000800000200000000000000000000000000000000000000000000000000000000000000006800000200000000000000000000000000000000000000000000000000001b91ca97c216a800000200000000000000000000000000000000000000000000000000000701c468c419e8000002000000000000000000000000000000000000000000000000000000efe3847eb84800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c4500000000000000000000000000000000000000000000000000008000002000000000000000000000000000000000000000000000000000000000000000128000002000000000000000000000000000000000000000000010ac16ce5f49333b50845c8000002000000000000000000000000000000000000000000001eee0f293b6dc405a7874800000200000000000000000000000000000000000000000000000000000000000000000"`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 3`] = `
[
  {
    "result": "Tether USD",
    "success": true,
  },
  {
    "result": "USDT",
    "success": true,
  },
  {
    "result": 6n,
    "success": true,
  },
  {
    "result": 485007730418026n,
    "success": true,
  },
  {
    "result": 123266744992158n,
    "success": true,
  },
  {
    "result": 16485028719492n,
    "success": true,
  },
  {
    "result": "Pendle",
    "success": true,
  },
  {
    "result": "PENDLE",
    "success": true,
  },
  {
    "result": 18n,
    "success": true,
  },
  {
    "result": 20155480847942246370149468n,
    "success": true,
  },
  {
    "result": 2336998592725751518886004n,
    "success": true,
  },
  {
    "result": 0n,
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 1`] = `
{
  "byteCode": "0x60046200004a60803960ff60e0600460846000600060046080600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d01608490036084f306fdde03",
  "splittedByteCodes": [
    "6004",
    "6200004a",
    "6080",
    "39",
    "60ff",
    "60e0",
    "6004",
    "6084",
    "6000",
    "6000",
    "6004",
    "6080",
    "6000",
    "73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6084",
    "90",
    "03",
    "6084",
    "f3",
    "",
    "",
    "06fdde03",
  ],
}
`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000a5465746865722055534400000000000000000000000000000000000000000000"`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 3`] = `
[
  {
    "result": "Tether USD",
    "success": true,
  },
]
`;
